src/ncp-spinel/SpinelNCPControlInterface.cpp

```diff
@@ -281,6 +281,42 @@ SpinelNCPControlInterface::remove_external_route(
 	return;
 }
 
+void SpinelNCPControlInterface::add_service(
+	uint32_t enterprise_number, 
+	const uint8_t *service_data, 
+	unsigned int service_data_len, 
+	bool stable, 
+	const uint8_t *server_data, 
+	unsigned int server_data_len, 
+	CallbackWithStatus cb
+) {
+	require_action(service_data != NULL, bail, cb(kWPANTUNDStatus_InvalidArgument));
+	require_action(service_data_len > 0, bail, cb(kWPANTUNDStatus_InvalidArgument));
+	require_action(server_data != NULL, bail, cb(kWPANTUNDStatus_InvalidArgument));
+	require_action(server_data_len > 0, bail, cb(kWPANTUNDStatus_InvalidArgument));
+	require_action(mNCPInstance->mEnabled, bail, cb(kWPANTUNDStatus_InvalidWhenDisabled));
+
+	mNCPInstance->add_service(enterprise_number, service_data, service_data_len, stable, server_data, server_data_len, cb);
+
```

Comment:
We can consider directly implementing the spinel exchange here from the `SpinelNCPControlInterface` (e.g. similar to `commissioner_add_joiner`). With this mode we can remove the `NCPInstanceBase` and `SpinelNCPInstance` related APIs (`add_service_on_ncp` and `add_service`) which would simplify the logic.

For addresses/routes/prefixes we go through the `NCPInstanceBase` class since it implements the mechanism to save the user-originated  entries (from `NCPInstanceBase-Addresses`)  and ensures to re-add them on NCP on NCP reset/reboot.

This PR does not currently save the "service" entries in `wpantund`. Not sure if we would need this in future. If we do, I'd say we should keep the current flow (go through the NCPInstanceBase), otherwise we can directly implement the logic in `SpinelNCPControlInterface`.

Thoughts?

Reply:
Think it could be advantageous to restore the service entries by saving it in wpantund. At least I have noticed most configuration is saved and restored, whether that is in wpantund or on the NCP flash itself. Otherwise we push the complexity in to the user's supervisor application (connected via DBus) to monitor resets and re-add services.

I will look in to adding this functionality.

Reply:
Done